### PR TITLE
feat: TOOLS-3077 - Add Server Diff Functionality to asconfig diff Command

### PR DESCRIPTION
**Pull request by a-spiker**
### **Summary**

This PR significantly enhances the `asconfig diff` command by:
- Adding support for comparing a local config file against a live Aerospike server (`diff server` subcommand).

---

#### **1. Feature: `diff server` Subcommand**
- Adds `asconfig diff server [flags] <path/to/config>` to compare a local config file (YAML or .conf) with the live configuration of a running Aerospike server.
- Adds `asconfig diff files [flags] <path/to/config1> <path/to/config2>` to compare two config file (YAML or .conf) 
- asconfig diff <path/to/config1> <path/to/config2> still is supported but throws a warning to use `asconfig diff files`
- Supports all necessary Aerospike connection flags.
- Provides clear help text and usage examples.
- Marks the feature as BETA and hides the `--format` flag from help output from diff server (auto-detected or settable for advanced use).

#### **2. CLI Help Improvements**
- Updates the top-level `diff` command help to mention both file-to-file and file-to-server comparison as subcommands.
- Clarifies argument requirements and format detection in both command and subcommand help texts.

#### **4. Test Coverage**
- Adds and refactors unit tests for:
  - Argument validation (too few/many args, non-existent files, etc.)
  - Format detection (flag vs. extension, invalid formats)
  - Edge cases for both file and server diff modes
  - Flat map diff logic (including logging enum and metadata handling)
- Ensures robust error handling and clear test descriptions.

#### **5. Refactoring and Logging**
- Refactors core diff logic into `runFileDiff` and `runServerDiff` for clarity and maintainability.
- Adds debug logging for key steps and error cases.

---

### **Motivation**

- Users need to compare local config files with the live state of Aerospike servers for troubleshooting and validation.
- Consistent and predictable format handling is critical for usability.

---

### **How to Use**

- **Compare two files: (Deprecated) for backward compatibility **
  ```
  asconfig diff file1.yaml file2.yaml
  ```
- **Compare two files:**
  ```
  asconfig diff files file1.yaml file2.yaml
  ```  
- **Compare a file with a running server:**
  ```
  asconfig diff server -h <host> -U <user> -P <password> file.conf
  ```
  (or with YAML: `asconfig diff server  ... file.yaml`)  

---

### **Notes**

- The `diff server` feature is marked as BETA as it uses generate functionality, which is in BETA
- All changes are backward compatible with previous `diff` usage.

---

### CODE REVIEW

File: cmd/diff.go [hunk 1]
```diff
@@ -167,10 +134,222 @@ func diffFlatMaps(m1 map[string]any, m2 map[string]any) []string {
...
 			res = append(res, fmt.Sprintf("%s:\n\t<: %v\n\t>: %v\n", k, v1, v2))
 		}
 	}
 
 	return res
 }
+
+// valuesEqual compares two values with type-aware comparison
+// This handles cases where values might be the same but different types
+// (e.g., int vs string, int vs int64, slices, etc.)
```
[the remark this answers is not in the document]
Comment (a-spiker) [reply cmd/diff.go:150]: Had assumed that slices would be in the same order during parsing. Made it order agnostic.

File: cmd/diff.go [hunk 2]
```diff
@@ -167,10 +134,222 @@ func diffFlatMaps(m1 map[string]any, m2 map[string]any) []string {
...
+}
+
+// isSlice checks if a value is a slice
+func isSlice(v any) bool {
+	str := fmt.Sprintf("%T", v)
+	return strings.HasPrefix(str, "[]")
+}
+
+// slicesEqual compares two values that might be slices
+func slicesEqual(v1, v2 any) bool {
```
[the remark this answers is not in the document]
Comment (a-spiker) [reply cmd/diff.go:192]: Had assumed that slices would be in the same order during parsing. Made it order agnostic.

File: cmd/diff.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (a-spiker) [reply cmd/diff.go:150]: https://go.dev/play/p/zmnP8QyevYl

During our parsing, if it fails.. thought of defensive programming.

**Review by dwelch-spike — CHANGES_REQUESTED** [review]
Overall this looks like a nice way to make comparing server configs easier 👍 I think we need to think about this feature some more before adding it though. Here are the 2 main things on my mind

1. I'm not sure about the type agnostic value equality checking code added. I think it might hide errors in the parsing code and potentially miss valid diffs, I'd hold off on adding code like that until a real issue is observed in the field or testing. I'm open to being convinced otherwise though

2. A nitpick, I don't like that the help output for `diff` includes all the cluster connection options by default now. When you are just diffing 2 files, none of them are needed.
See below
```
./asconfig diff --help
Diff is used to compare differences between Aerospike configuration files.
                                It is used on two files of the same format from any format
                                supported by the asconfig tool, e.g. yaml or Aerospike config.
                                Schema validation is not performed on either file. The file names must end with
                                extensions signifying their formats, e.g. .conf or .yaml, or --format must be used.

                                When using --server flag, compare a local configuration file against
                                the live configuration from a running Aerospike server. In this mode,
                                only one config file path is required as an argument.

                                Examples:
                                  # Compare two local configuration files
                                  asconfig diff aerospike1.yaml aerospike2.yaml
                                  
                                  # Compare local config against running server
                                  asconfig diff --server -h 127.0.0.1 -U admin -P admin aerospike.yaml
                                  
                                  # Compare local config against server with TLS
                                  asconfig diff -s -h aerospike.example.com --tls-enable aerospike.conf

Usage:
  asconfig diff [flags] <path/to/config1> <path/to/config2>

Flags:
      --auth INTERNAL,EXTERNAL,PKI                                                                  The authentication mode used by the Aerospike server. INTERNAL 
                                                                                                    uses standard user/pass. EXTERNAL uses external methods (like LDAP) 
                                                                                                    which are configured on the server. EXTERNAL requires TLS. PKI allows 
                                                                                                    TLS authentication and authorization based on a certificate. No 
                                                                                                    username needs to be configured. (default INTERNAL)
  -F, --format string                                                                               The format of the source file(s). Valid options are: yaml, yml, and conf. (default "conf")
  -h, --host host[:tls-name][:port][,...]                                                           The Aerospike host. (default 127.0.0.1)
  -P, --password "env-b64:<env-var>,b64:<b64-pass>,file:<pass-file>,<clear-pass>"                   The Aerospike password to use to connect to the Aerospike 
                                                                                                    cluster.
  -p, --port int                                                                                    The default Aerospike port. (default 3000)
  -s, --server                                                                                      Compare local config file against a running Aerospike server configuration.
      --tls-cafile env-b64:<cert>,b64:<cert>,<cert-file-name>                                       The CA used when connecting to Aerospike.
      --tls-capath <cert-path-name>                                                                 A path containing CAs for connecting to Aerospike.
      --tls-certfile env-b64:<cert>,b64:<cert>,<cert-file-name>                                     The certificate file for mutual TLS authentication with 
                                                                                                    Aerospike.
      --tls-enable                                                                                  Enable TLS authentication with Aerospike. If false, other TLS 
                                                                                                    options are ignored.
      --tls-keyfile env-b64:<cert>,b64:<cert>,<cert-file-name>                                      The key file used for mutual TLS authentication with Aerospike.
      --tls-keyfile-password "env-b64:<env-var>,b64:<b64-pass>,file:<pass-file>,<clear-pass>"       The password used to decrypt the key file if encrypted.
      --tls-name string                                                                             The server TLS context to use to authenticate the connection to 
                                                                                                    Aerospike.
      --tls-protocols "[[+][-]all] [[+][-]TLSv1] [[+][-]TLSv1.1] [[+][-]TLSv1.2] [[+][-]TLSv1.3]"   Set the TLS protocol selection criteria. This format is the same 
                                                                                                    as Apache's SSLProtocol documented at 
                                                                                                    https://httpd.apache.org/docs/current/mod/mod_ssl.html#ssl protocol. (default +TLSv1.2)
  -U, --user string                                                                                 The Aerospike user to use to connect to the Aerospike cluster.

Global Flags:
      --config-file string   Config file (default is /etc/aerospike/astools)
  -u, --help                 Display help information
      --instance string      For support of the aerospike tools toml schema. Sections with the 
                             instance are read. e.g in the case where instance 'a' is specified 
                             sections 'cluster_a', 'uda_a' are read.
  -l, --log-level string     Set the logging detail level. Valid levels are: [panic fatal error warning info debug trace] (default "info")
  -V, --version              Display version.
```
I think you can make it cleaner by adding the cluster flags to their own group with something like this in the newDiffCmd function
```
	// Group server-specific flags for cleaner help output
	res.SetHelpFunc(func(cmd *cobra.Command, args []string) {
		// Group flags for cleaner help output
		fg := cmd.Flags()
		serverFg := pflag.NewFlagSet("server", pflag.ContinueOnError)
		serverFg.AddFlagSet(asFlagSet)

		// Create dedicated flag groups
		requiredFlags := pflag.NewFlagSet("required", pflag.ContinueOnError)
		requiredFlags.AddFlag(fg.Lookup("format"))
		requiredFlags.AddFlag(fg.Lookup("server"))

		// Print default help, but with custom flag groupings
		fmt.Fprintf(cmd.OutOrStdout(), "%s\n\n", cmd.Long)
		fmt.Fprintf(cmd.OutOrStdout(), "Usage:\n  %s\n\n", cmd.Use)
		fmt.Fprintf(cmd.OutOrStdout(), "Flags:\n%s\n", requiredFlags.FlagUsages())
		fmt.Fprintf(cmd.OutOrStdout(), "Server Connection Flags (for use with --server):\n%s\n", serverFg.FlagUsages())
		fmt.Fprintf(cmd.OutOrStdout(), "Global Flags:\n%s\n", cmd.InheritedFlags().FlagUsages())
	})
```

**Comment by dwelch-spike** [comment]
I played around with making the `--server` flag a subcommand that would be used like `asconfig diff server [flags] <local/path>` and the help looks pretty clean. I think this would be a clean way to organize it. @a-spiker what do you think?

```
 ./asconfig diff --help
Diff is used to compare differences between Aerospike configuration files.
                                It is used on two files of the same format from any format
                                supported by the asconfig tool, e.g. yaml or Aerospike config.
                                Schema validation is not performed on either file. The file names must end with
                                extensions signifying their formats, e.g. .conf or .yaml, or --format must be used.

                                To compare a local file against a running server, see the 'server' subcommand.

                                Examples:
                                  # Compare two local configuration files
                                  asconfig diff aerospike1.yaml aerospike2.yaml

Usage:
  asconfig diff <path/to/config1> <path/to/config2> [flags]
  asconfig diff [command]

Available Commands:
  server      Compare a local config file against a running Aerospike server.

Flags:
  -F, --format string   The format of the source file(s). Valid options are: yaml, yml, and conf. (default "conf")

Global Flags:
      --config-file string   Config file (default is /etc/aerospike/astools)
  -u, --help                 Display help information
      --instance string      For support of the aerospike tools toml schema. Sections with the 
                             instance are read. e.g in the case where instance 'a' is specified 
                             sections 'cluster_a', 'uda_a' are read.
  -l, --log-level string     Set the logging detail level. Valid levels are: [panic fatal error warning info debug trace] (default "info")
  -V, --version              Display version.

Use "asconfig diff [command] --help" for more information about a command.
```

File: cmd/diff.go [hunk 3]
```diff
@@ -167,10 +135,287 @@ func diffFlatMaps(m1 map[string]any, m2 map[string]any) []string {
...
+// (e.g., int vs string, int vs int64, slices, etc.)
+func valuesEqual(v1, v2 any) bool {
+	// Check if either value is a slice first (to avoid panic on direct comparison)
+	if isSlice(v1) || isSlice(v2) {
+		return slicesEqual(v1, v2)
+	}
+
+	// Convert both to strings and compare first
+	str1 := fmt.Sprintf("%v", v1)
+	str2 := fmt.Sprintf("%v", v2)
```
[the remark this answers is not in the document]
Comment (a-spiker) [reply cmd/diff.go]: Removed the extra logic around comparing

File: cmd/diff.go [hunk 4]
```diff
@@ -167,10 +135,287 @@ func diffFlatMaps(m1 map[string]any, m2 map[string]any) []string {
...
+// convertToStringSlice converts any slice type to a slice of strings for comparison
+// Uses reflection for efficient direct access to slice elements
+func convertToStringSlice(v any) []string {
+	if v == nil {
+		return []string{}
+	}
+
+	// Use reflection to check if it's actually a slice
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Slice {
```
[the remark this answers is not in the document]
Comment (a-spiker) [reply cmd/diff.go]: Removed the extra logic around comparing

**Review by dwelch-spike — DISMISSED** [review]
LGTM I'm happy with the `diff files` and `diff server` subcommand split.

File: cmd/diff.go [hunk 5]
```diff
@@ -32,95 +40,253 @@ var diffCmd = newDiffCmd()
...
+		Long: `BETA: Diff is used to compare a local configuration file against the configuration of a running Aerospike server. 
+				In this mode, only one config file path is required as an argument.
+				Note: The configuration file can be in yaml or conf format.`,
+		Example: `Diff a local .conf file against a running server
+  				asconfig diff server -h 127.0.0.1:3000 aerospike.conf`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			logger.Debug("Running server diff command")
+			return runServerDiff(cmd, args)
+		},
+	}
```
Comment (dwelch-spike) [inline cmd/diff.go:109]:
```suggestion
func newDiffServerCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "server [flags] <path/to/config>",
		Short: "BETA: Diff a local config file against a running Aerospike server's configuration.",
		Long: `BETA: Diff is used to compare a local configuration file against the configuration of a running Aerospike server. 
				This is useful for spotting drift between expected and actual Aerospike server configurations.
				In this mode, only one config file path is required as an argument.
				Note: The configuration file can be in yaml or conf format.`,
		Example: `Diff a local .conf file against a running server
  				asconfig diff server -h 127.0.0.1:3000 aerospike.conf`,
		RunE: func(cmd *cobra.Command, args []string) error {
			logger.Debug("Running server diff command")
			return runServerDiff(cmd, args)
		},
	}
```
A little note about how to use it might be good. Use this if you want.

**Review by dwelch-spike — APPROVED** [review]
LGTM!

**Review by tanmayja — APPROVED** [review]
